File: SQL Server 2012/Exercises/Chapter 20/C#/Exercise 20-02/MusicStoreImageManager/ProductDB.cs
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using System.IO;
using System.Windows.Forms;

namespace MusicStoreImageManager
{
    class ProductDB
    {
        // define the directory for the images
        static string imagesPath = "C:/Murach/SQL Server 2012/Images/";

        public static SqlConnection GetConnection()
        {
            SqlConnection connection = new SqlConnection();
            connection.ConnectionString = 
                "Data Source=localhost\\SqlExpress;" + 
                "Initial Catalog=MusicStore;Integrated Security=True";
            return connection;
        }

        public static void WriteImage(int productID, string imageName)
        {
            SqlConnection connection = null;
            SqlTransaction transaction = null;
            try
            {
                // 1. Set up the input stream from the image file
                string filepath = imagesPath + imageName;
                if (File.Exists(filepath) == false)
                    throw new Exception("File Not Found: " + filepath);

                FileStream sourceStream = new FileStream(
                    filepath,
                    FileMode.Open,
                    FileAccess.Read);

                // 2. Initialize the row in the table
                connection = GetConnection();

                SqlCommand command = new SqlCommand();
                command.Connection = connection;
                command.CommandText =
                    "INSERT INTO ProductImages " +
                    "VALUES (@ProductID, " +
                    "        CAST(@RowID AS uniqueidentifier), 0)";

                Guid rowID = Guid.NewGuid();
                command.Parameters.AddWithValue("@ProductID", productID);
                command.Parameters.AddWithValue("@RowID", rowID);

                connection.Open();
                command.ExecuteNonQuery();

                // 3. Get a reference to the BLOB
                transaction = connection.BeginTransaction();
                command.Transaction = transaction;
                command.CommandText =
                    "SELECT ProductImage.PathName(), " +
                    "       GET_FILESTREAM_TRANSACTION_CONTEXT() " +
                    "FROM ProductImages " +
                    "WHERE RowID = CAST(@RowID AS uniqueidentifier)";
                command.Parameters.Clear();
                command.Parameters.AddWithValue("@RowID", rowID);

                SqlDataReader reader = command.ExecuteReader();
                if (reader.Read() == false)
                {
                    throw new Exception("Unable to get path and context for BLOB.");
                }
                string path = (string)reader[0];
                byte[] context = (byte[])reader[1];
                reader.Close();

                // 4. Set up the output stream to the database
                SqlFileStream targetStream = 
                    new SqlFileStream(path, context, FileAccess.Write);

                // 5. Read from file and write to database
                int blockSize = 1024 * 512;
                byte[] buffer = new byte[blockSize];
                int bytesRead = sourceStream.Read(buffer, 0, buffer.Length);
                while (bytesRead > 0)
                {
                    targetStream.Write(buffer, 0, bytesRead);
                    bytesRead = sourceStream.Read(buffer, 0, buffer.Length);
                }

                targetStream.Close();
                sourceStream.Close();
                transaction.Commit();
            }
            catch (Exception e)
            {
                if (transaction != null)
                    transaction.Rollback();
                throw e;
            }
            finally
            {
                if (connection != null)
                    connection.Close();
            }
        }

        public static void ReadImage(int imageID, PictureBox pictureBox)
        {
            SqlConnection connection = null;
            SqlTransaction transaction = null;
            try
            {
                connection = GetConnection();
                connection.Open();
                transaction = connection.BeginTransaction();

                SqlCommand command = new SqlCommand();
                command.Connection = connection;
                command.Transaction = transaction;
                command.CommandText =
                    "SELECT ProductImage.PathName(), " +
                    "       GET_FILESTREAM_TRANSACTION_CONTEXT() " +
                    "FROM ProductImages " +
                    "WHERE ImageID = @ImageID";
                command.Parameters.AddWithValue("@ImageID", imageID);

                SqlDataReader reader = command.ExecuteReader();
                if (reader.Read() == false)
                {
                    throw new Exception("Unable to get path and context for BLOB.");
                }
                string path = (string)reader[0];
                byte[] context = (byte[])reader[1];
                reader.Close();

                // Set up the input stream from the database
                SqlFileStream sourceStream =
                    new SqlFileStream(path, context, FileAccess.Read);

                // Set up the output stream to the picture box
                MemoryStream targetStream = new MemoryStream();

                // Stream the data from the database to the form
                int blockSize = 1024 * 512;
                byte[] buffer = new byte[blockSize];
                int bytesRead = sourceStream.Read(buffer, 0, buffer.Length);
                while (bytesRead > 0)
                {
                    targetStream.Write(buffer, 0, buffer.Length);
                    bytesRead = sourceStream.Read(buffer, 0, buffer.Length);
                }
                pictureBox.Image = System.Drawing.Image.FromStream(targetStream);
                sourceStream.Close();
                targetStream.Close();
            }
            catch (Exception e)
            {
                throw e;
            }
            finally
            {
                if (connection != null)
                    connection.Close();
            }
        }

        public static List<int> GetImageIDList()
        {
            SqlConnection connection = null;
            try
            {
                connection = GetConnection();

                SqlCommand command = new SqlCommand();
                command.Connection = connection;
                command.CommandText = 
                    "SELECT ImageID FROM ProductImages " +
                    "ORDER BY ImageID";

                connection.Open();
                SqlDataReader reader = command.ExecuteReader();
 
                List<int> imageIDList = new List<int>();
                while (reader.Read())
                {
                    int imageID = (int)reader[0];
                    imageIDList.Add(imageID);
                }
                reader.Close();

                return imageIDList;
            }
            catch (Exception e)
            {
                throw e;
            }
            finally
            {
                if (connection != null)
                    connection.Close();
            }
        }
    }
}

File: SQL Server 2012/Exercises/Chapter 20/VB/Chapter 20-02/MusicStoreImageManager/ProductDB.vb
Imports System.Data.SqlClient
Imports System.Data.SqlTypes
Imports System.IO

Public Class ProductDB

    ' The directory for the images
    Shared imagesPath As String = "C:/Murach/SQL Server 2012/Images/"

    Protected Shared Function GetConnection() As SqlConnection
        Dim connection As SqlConnection = New SqlConnection()
        connection.ConnectionString =
            "Data Source=localhost\SqlExpress;" &
            "Initial Catalog=MusicStore;Integrated Security=True"
        Return connection
    End Function

    Public Shared Sub WriteImage(ByVal productID As Integer, ByVal imageName As String)
        Dim connection As SqlConnection = Nothing
        Dim transaction As SqlTransaction = Nothing
        Try
            ' 1. Set up the input stream from the image file
            Dim filepath As String = imagesPath + imageName
            If File.Exists(filepath) = False Then
                Throw New Exception("File Not Found: " + filepath)
            End If

            Dim sourceStream As FileStream = New FileStream(
                filepath,
                FileMode.Open,
                FileAccess.Read)

            ' 2. Initialize the row in the table
            connection = GetConnection()

            Dim command As SqlCommand = New SqlCommand()
            command.Connection = connection
            command.CommandText =
                "INSERT INTO ProductImages " &
                "VALUES (@ProductID, " &
                "        CAST(@RowID AS uniqueidentifier), 0)"

            Dim rowID As Guid = Guid.NewGuid()
            command.Parameters.AddWithValue("@ProductID", productID)
            command.Parameters.AddWithValue("@RowID", rowID)

            connection.Open()
            command.ExecuteNonQuery()

            ' 3. Get a reference to the BLOB
            transaction = connection.BeginTransaction()
            command.Transaction = transaction
            command.CommandText = _
                "SELECT ProductImage.PathName(), " & _
                "       GET_FILESTREAM_TRANSACTION_CONTEXT() " & _
                "FROM ProductImages " & _
                "WHERE RowID = CAST(@RowID AS uniqueidentifier)"
            command.Parameters.Clear()
            command.Parameters.AddWithValue("@RowID", rowID)

            Dim reader As SqlDataReader = command.ExecuteReader()
            If reader.Read() = False Then
                Throw New Exception("Unable to get path and context for BLOB.")
            End If
            Dim path As String = reader(0)
            Dim context As Byte() = reader(1)
            reader.Close()

            ' 4. Set up the output stream to the database
            Dim targetStream As SqlFileStream = _
                New SqlFileStream(path, context, FileAccess.Write)

            ' 5. Read from file and write to database
            Dim blockSize As Integer = 1024 * 512
            Dim buffer(blockSize) As Byte

            Dim bytesRead As Integer = sourceStream.Read(buffer, 0, buffer.Length)
            While bytesRead > 0
                targetStream.Write(buffer, 0, bytesRead)
                bytesRead = sourceStream.Read(buffer, 0, buffer.Length)
            End While

            targetStream.Close()
            sourceStream.Close()
            transaction.Commit()
        Catch e As Exception
            If Not IsDBNull(transaction) Then
                transaction.Rollback()
            End If
            Throw e
        Finally
            If Not IsDBNull(connection) Then
                connection.Close()
            End If
        End Try
    End Sub


    Public Shared Sub ReadImage(ByVal imageID As Integer,
        ByVal pictureBox As PictureBox)

        Dim connection As SqlConnection = Nothing
        Dim transaction As SqlTransaction = Nothing
        Try
            connection = GetConnection()
            connection.Open()
            transaction = connection.BeginTransaction()

            Dim command As SqlCommand = New SqlCommand()
            command.Connection = connection
            command.Transaction = transaction
            command.CommandText = _
                "SELECT ProductImage.PathName(), " & _
                "       GET_FILESTREAM_TRANSACTION_CONTEXT() " & _
                "FROM ProductImages " & _
                "WHERE ImageID = @ImageID"
            command.Parameters.AddWithValue("@ImageID", imageID)

            Dim reader As SqlDataReader = command.ExecuteReader()
            If reader.Read() = False Then
                Throw New Exception("Unable to get path and context for BLOB.")
            End If
            Dim path As String = reader(0)
            Dim context As Byte() = reader(1)
            reader.Close()

            ' Set up the input stream from the database
            Dim sourceStream As SqlFileStream = _
                New SqlFileStream(path, context, FileAccess.Read)

            ' Set up the output stream to the picture box
            Dim targetStream As MemoryStream = New MemoryStream()

            Dim blockSize As Integer = 1024 * 512
            Dim buffer(blockSize) As Byte
            Dim imageBytes As List(Of Byte) = New List(Of Byte)
            Dim bytesRead As Integer = sourceStream.Read(buffer, 0, buffer.Length)
            While (bytesRead > 0)
                targetStream.Write(buffer, 0, buffer.Length)
                bytesRead = sourceStream.Read(buffer, 0, buffer.Length)
            End While
            pictureBox.Image = System.Drawing.Image.FromStream(targetStream)
            sourceStream.Close()
            targetStream.Close()
        Catch e As Exception
            Throw e
        Finally
            If Not IsDBNull(connection) Then
                connection.Close()
            End If
        End Try
    End Sub


    Public Shared Function GetImageIDList() As List(Of Integer)
        Dim connection As SqlConnection = Nothing
        Try
            connection = GetConnection()

            Dim command As SqlCommand = New SqlCommand()
            command.Connection = connection
            command.CommandText = _
                "SELECT ImageID FROM ProductImages " & _
                "ORDER BY ImageID"

            connection.Open()
            Dim reader As SqlDataReader = command.ExecuteReader()

            Dim imageIDList As List(Of Integer) = New List(Of Integer)
            While reader.Read()
                Dim imageID As Integer = reader(0)
                imageIDList.Add(imageID)
            End While
            reader.Close()

            Return imageIDList
        Catch e As Exception
            Throw e
        Finally
            If Not IsDBNull(connection) Then
                connection.Close()
            End If
        End Try
    End Function

End Class
